From 9681a3d12be2f27d90c3392e22a37d9c8769686a Mon Sep 17 00:00:00 2001
From: Karthik Poosa <kpoosa2@codeaurora.org>
Date: Wed, 17 Jul 2019 11:59:30 +0530
Subject: [PATCH] NFC: Fix device bootup issue

NFCC device probe was getting blocked as it was
waiting for interrupt from NFCC.

Added timeout waiting for interrupt during device probe.

Change-Id: I3685615b108e2535cd0e62934b4e7a5bc5d742fe
Signed-off-by: Karthik Poosa <kpoosa2@codeaurora.org>
---
 drivers/nfc/nq-nci.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/nfc/nq-nci.c b/drivers/nfc/nq-nci.c
index b56d049cbb74..2c8f425ccb1e 100644
--- a/drivers/nfc/nq-nci.c
+++ b/drivers/nfc/nq-nci.c
@@ -21,6 +21,7 @@
 #ifdef CONFIG_COMPAT
 #include <linux/compat.h>
 #endif
+#include <linux/jiffies.h>
 
 struct nqx_platform_data {
 	unsigned int irq_gpio;
@@ -51,6 +52,7 @@ MODULE_DEVICE_TABLE(of, msm_match_table);
 #define NCI_RESET_NTF_LEN		13
 #define NCI_GET_VERSION_CMD_LEN		8
 #define NCI_GET_VERSION_RSP_LEN		12
+#define MAX_IRQ_WAIT_TIME		(90)	//in ms
 
 struct nqx_dev {
 	wait_queue_head_t	read_wq;
@@ -896,8 +898,9 @@ static int nfcc_hw_check(struct i2c_client *client, struct nqx_dev *nqx_dev)
 		goto err_nfcc_reset_failed;
 	}
 	nqx_enable_irq(nqx_dev);
-	ret = wait_event_interruptible(nqx_dev->read_wq, !nqx_dev->irq_enabled);
-	if (ret < 0) {
+	ret = wait_event_interruptible_timeout(nqx_dev->read_wq,
+		!nqx_dev->irq_enabled, msecs_to_jiffies(MAX_IRQ_WAIT_TIME));
+	if (ret <= 0) {
 		nqx_disable_irq(nqx_dev);
 		goto err_nfcc_hw_check;
 	}
@@ -913,8 +916,9 @@ static int nfcc_hw_check(struct i2c_client *client, struct nqx_dev *nqx_dev)
 		goto err_nfcc_hw_check;
 	}
 	nqx_enable_irq(nqx_dev);
-	ret = wait_event_interruptible(nqx_dev->read_wq, !nqx_dev->irq_enabled);
-	if (ret < 0) {
+	ret = wait_event_interruptible_timeout(nqx_dev->read_wq,
+		!nqx_dev->irq_enabled, msecs_to_jiffies(MAX_IRQ_WAIT_TIME));
+	if (ret <= 0) {
 		nqx_disable_irq(nqx_dev);
 		goto err_nfcc_hw_check;
 	}
-- 
2.53.0

